ci/remote-watch.py

```diff
@@ -0,0 +1,201 @@
+    if ci == GITHUB:
+        return os.environ["GITHUB_EVENT_NAME"]
+    elif ci == TRAVIS:
+        return os.environ["TRAVIS_EVENT_TYPE"]
+    return None
+
+
+def get_repo_slug():
+    ci = get_current_ci()
+    if ci == GITHUB:
+        return os.environ["GITHUB_REPOSITORY"]
+    elif ci == TRAVIS:
+        return os.environ["TRAVIS_REPO_SLUG"]
+    return None
+
+
+def git_branch_info():
+    ref = None
+    try:
+        head = git("symbolic-ref", "-q", "HEAD")
+        ref = git("for-each-ref", "--format=%(upstream:short)", head)
+    except subprocess.CalledProcessError:
+        pass
+    remote = None
+    expected_sha = None
+    if ref:
+        (remote, ref) = ref.split("/", 1)
+        ref = "refs/heads/" + ref
+    else:
+        remote = git_remotes()[0]
+        ref = os.getenv("TRAVIS_PULL_REQUEST_BRANCH")
+        if ref:
+            TRAVIS_PULL_REQUEST = os.environ["TRAVIS_PULL_REQUEST"]
+            if os.getenv("TRAVIS_EVENT_TYPE") == "pull_request":
+                ref = "refs/pull/{}/merge".format(TRAVIS_PULL_REQUEST)
+            else:
+                ref = "refs/heads/{}".format(TRAVIS_PULL_REQUEST)
+            expected_sha = os.getenv("TRAVIS_COMMIT")
+        else:
+            ref = os.getenv("CI_REF")
+    if not remote:
+        raise ValueError("Invalid remote: {!r}".format(remote))
+    if not ref:
+        raise ValueError("Invalid ref: {!r}".format(ref))
+    expected_sha = git("rev-parse", "--verify", "HEAD")
+    return (ref, remote, expected_sha)
+
+
+def terminate_my_process_group():
+    result = 0
+    timeout = 15
+    try:
+        logger.warning("Attempting kill...")
+        if sys.platform == "win32":
+            os.kill(0, signal.CTRL_BREAK_EVENT)  # this might get ignored
+            time.sleep(timeout)
+            os.kill(os.getppid(), signal.SIGTERM)
+        else:
+            os.kill(os.getppid(), signal.SIGTERM)  # apparently this is needed
+            time.sleep(timeout)
+            os.kill(0, signal.SIGKILL)
+    except OSError as ex:
+        if ex.errno not in (errno.EBADF, errno.ESRCH):
+            raise
+        logger.error("Kill error %s: %s", ex.errno, ex.strerror)
+        result = ex.errno
+    return result
+
+
+def yield_poll_schedule():
+    schedule = [0, 5, 5, 10, 20, 40, 40] + [60] * 5 + [120] * 10 + [300, 600]
+    for item in schedule:
+        yield item
+    while True:
+        yield schedule[-1]
+
+
+def monitor():
+    ci = get_current_ci() or ""
+    (ref, remote, expected_sha) = git_branch_info()
+    expected_line = "{}\t{}".format(expected_sha, ref)
+    commit_msg = git("show", "-s", "--format=%B", "HEAD^-")
+    keep_alive = False
+    for match in re.findall("(?m)^([# ]*CI_KEEP_ALIVE(:(.*))?)$", commit_msg):
+        if not match[2].strip():
+            keep_alive = True
+        else:
+            for ci_name in match[2].split(","):
+                if ci_name.strip().lower() == ci.lower():
+                    keep_alive = True
```

Comment:
if keep_alive is true, we should just exit this script and let the CI run, stop polling git commits.

Reply:
Huh, good point. I can't recall why I did it this way originally.